src/ui/public/persisted_state/persisted_state.js

```diff
@@ -0,0 +1,194 @@
+    };
+
+    PersistedState.prototype.reset = function (key) {
+      this.set(key, undefined);
+    };
+
+    PersistedState.prototype.clear = function (key) {
+      this.set(key, null);
+    };
+
+    PersistedState.prototype.createChild = function (path, value) {
+      return new PersistedState(value, this._getIndex(path), this._parent || this);
+    };
+
+    PersistedState.prototype.getChanges = function () {
+      return _.cloneDeep(this._changedState);
+    };
+
+    PersistedState.prototype.toJSON = function () {
+      return this.get();
+    };
+
+    PersistedState.prototype.toString = function () {
+      return JSON.stringify(this.toJSON());
+    };
+
+    PersistedState.prototype.fromString = function (input) {
+      return this.set(JSON.parse(input));
+    };
+
+    PersistedState.prototype._getIndex = function (key) {
+      if (_.isUndefined(key)) return this._path;
+      return (this._path || []).concat(toPath(key));
+    };
+
+    PersistedState.prototype._getDefault = function () {
+      var def = (this._hasPath()) ? undefined : {};
+      return (this._parent ? this.get() : def);
+    };
+
+    PersistedState.prototype._setPath = function (path) {
+      var isString = _.isString(path);
+      var isArray = _.isArray(path);
+
+      if (!isString && !isArray) return [];
+      return (isString) ? [this._getIndex(path)] : path;
+    };
+
+    PersistedState.prototype._hasPath = function () {
+      return this._path.length > 0;
+    };
+
+    PersistedState.prototype._get = function (key, def) {
+      // delegate to parent instance
+      if (this._parent) return this._parent._get(this._getIndex(key), key);
+
+      // no path and no key, get the whole state
+      if (!this._hasPath() && _.isUndefined(key)) {
+        return this._mergedState;
+      }
+
+      return _.get(this._mergedState, this._getIndex(key), def);
+    };
+
+    PersistedState.prototype._set = function (key, value, initialChildState, defaultChildState) {
+      var self = this;
+      var initialState = !this._initialized;
+      var keyPath = this._getIndex(key);
+
+      // if this is the initial state value, save value as the default
+      if (initialState) {
+        this._changedState = {};
+        if (!this._hasPath() && _.isUndefined(key)) this._defaultState = value;
+        else this._defaultState = _.set({}, keyPath, value);
+      }
+
+      // delegate to parent instance, passing child's default value
+      if (this._parent) {
+        return this._parent._set(keyPath, value, initialState, this._defaultState);
+      }
+
+      // everything in here affects only the parent state
+      if (!initialState) {
+        // no path and no key, set the whole state
+        if (!this._hasPath() && _.isUndefined(key)) {
+          // check for changes and emit an event when found
+          if (!_.isEqual(this._changedState, value)) this.emit('change');
+          if (!initialChildState) this._changedState = value;
```

Comment:
I think we need to set _changedState before we emit change, or else listeners can't read the changed value